src/encrypt.rs

```diff
@@ -0,0 +1,464 @@
+        serde_cbor::to_vec(self)
+    }
+}
+
+///  Implementation of the COSE_Encrypt0 structure as defined in
+///  [RFC8152](https://tools.ietf.org/html/rfc8152#section-5.2).
+///
+///  The COSE_Encrypt0 encrypted structure does not have the ability to
+///  specify recipients of the message.  The structure assumes that the
+///  recipient of the object will already know the identity of the key to
+///  be used in order to decrypt the message.  If a key needs to be
+///  identified to the recipient, the enveloped structure ought to be
+///  used.
+///
+///  The COSE_Encrypt0 structure can be encoded as either tagged or
+///  untagged depending on the context it will be used in.  A tagged
+///  COSE_Encrypt0 structure is identified by the CBOR tag 16.  The CDDL
+///  fragment that represents this is:
+///
+///  COSE_Encrypt0_Tagged = #6.16(COSE_Encrypt0)
+///
+///  The COSE_Encrypt0 structure is a CBOR array.  The fields of the array in
+///  order are:
+///
+///  protected:  This is as described in Section 3.
+///
+///  unprotected:  This is as described in Section 3.
+///
+///  ciphertext:  This is as described in Section 5.1.
+///
+///  The CDDL fragment that represents the above text for COSE_Encrypt0
+///  follows.
+///
+///  COSE_Encrypt0 = [
+///      Headers,
+///      ciphertext : bstr / nil,
+///  ]
+#[derive(Debug, Clone, Deserialize)]
+pub struct COSEEncrypt0 {
+    /// protected: empty_or_serialized_map,
+    protected: ByteBuf,
+    /// unprotected: HeaderMap
+    unprotected: HeaderMap,
+    /// ciphertext: bstr
+    /// The spec allows ciphertext to be nil and transported separately, but it's not useful at the
+    /// moment, so this is just a ByteBuf for simplicity.
+    ciphertext: ByteBuf,
+}
+
+impl Serialize for COSEEncrypt0 {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut seq = serializer.serialize_seq(Some(3))?;
+        seq.serialize_element(&self.protected)?;
+        seq.serialize_element(&self.unprotected)?;
+        seq.serialize_element(&self.ciphertext)?;
+        seq.end()
+    }
+}
+
+impl COSEEncrypt0 {
+    /// Creates a new instance of the COSE_Encrypt0 structure and encrypts the provided payload.
+    /// https://datatracker.ietf.org/doc/html/rfc8152#section-5.3
+    pub fn new(payload: &[u8], cipher: Cipher, key: &[u8]) -> Result<Self, COSEError> {
+        let mut iv = vec![0; cipher.iv_len().unwrap()];
+        rand_bytes(&mut iv).unwrap();
+
+        let cose_alg = match COSEAlgorithm::from_nid(cipher.nid()) {
+            None => {
+                return Err(COSEError::UnsupportedError(
+                    "Unsupported encryption algorithm".to_string(),
+                ))
+            }
+            Some(v) => v,
+        };
+        let cose_alg_value = cose_alg.value();
+        let mut protected = HeaderMap::new();
+        protected.insert(1.into(), CborValue::Integer(cose_alg_value as i128));
```

Comment:
A `const KTY` with a comment about expected type would be very helpful (or an enum).